sky/optimizer.py

```diff
@@ -139,8 +145,84 @@ def optimize(dag: 'dag_lib.Dag',
                 # Make sure to remove the dummy source/sink nodes, even if the
                 # optimization fails.
                 Optimizer._remove_dummy_source_sink_nodes(dag)
+                Optimizer._remove_storage_nodes_and_move_to_edge_attributes(
+                    dag, storage_nodes_info)
             return dag
 
+    @staticmethod
+    def _remove_storage_nodes_and_move_to_edge_attributes(
+        dag: 'dag_lib.Dag',
+        edges_to_add: List[Tuple[task_lib.Task, task_lib.Task, task_lib.Task,
+                                 TaskData]]):
+        """Removes the storage nodes and adds the storage information to the
+        edges."""
+        with dag:
+            for src, storage_node, dst, data in edges_to_add:
+                dag.remove_edge(src, storage_node)
+                dag.remove_edge(storage_node, dst)
+                dag.remove(storage_node)
+                task_edge = dag.add_edge(src, dst)
+                task_edge.with_data(source_path=data.source_path,
+                                    target_path=data.target_path,
+                                    size_gb=data.size_gb)
+
+                # TODO(wenjie): support r2 storage.
+                if storage_node.best_resources is not None:
+                    assert storage_node.best_resources.cloud is not None
+                    storage_type = StoreType.from_cloud(
+                        str(storage_node.best_resources.cloud))
+                    task_edge.best_storage = (
+                        storage_type, storage_node.best_resources.region)
+
+    @staticmethod
+    def _add_storage_nodes_for_data_transfer(
+        dag: 'dag_lib.Dag'
+    ) -> List[Tuple[task_lib.Task, task_lib.Task, task_lib.Task, TaskData]]:
+        """Adds special nodes for storage buckets. Specifically, it adds dummy
+        nodes between nodes that are connected by with_data edges."""
+        graph = dag.get_graph()
+        edges_to_add = []
+
+        # TODO(wenjie): this is a hacky way to use the compute resources to
+        # represent the storage resources. We should have a separate class for
+        # storage resources. Also, we should add the logic of selecting
+        # available bucket resources before optimizing the DAG.
+        for src, dst, edge_data in graph.edges(data=True):
+            if edge_data['edge'].data is not None:
+                storage_node = task_lib.Task(
+                    f'{src.name}_to_{dst.name}_storage')
+                data = edge_data['edge'].data
+                storage_node.set_resources({
+                    resources_lib.Resources(clouds.AWS()),
+                    resources_lib.Resources(clouds.GCP()),
+                    resources_lib.Resources(clouds.Azure()),
+                    resources_lib.Resources(clouds.IBM())
+                })
```

Comment:
pls checkout this. lets not add clouds that the user not enabled to save some time at optimization :)

https://github.com/skypilot-org/skypilot/blob/334b268ce27bb9a32a452959648fe98e4307dd9b/sky/data/storage.py#L76-L93

Reply:
also remove R2